branchpro/apps/_simulation.py

```
@@ -0,0 +1,121 @@
+
+        Parameters
+        ----------
+        simulator
+            (SimulatorController) a BranchPro model and the time bounds
+            between which you run the simulator.
+        """
+        if not issubclass(type(simulator), bp.SimulationController):
+            raise TypeError('Simulatior needs to be a SimulationController')
+
+        model = simulator.model
+
+        if not issubclass(type(model), bp.BranchProModel):
+            raise TypeError('Models needs to be a BranchPro')
+
+        bounds = simulator.get_time_bounds()
+        mid_point = sum(bounds)/2
+
+        self.sliders.add_slider(
+            'Initial Cases', 'init_cond', 10.0, 0.0, 100.0, 10.0)
+        self.sliders.add_slider('Initial R', 'r0', 2.0, 0.1, 10.0, 0.01)
+        self.sliders.add_slider('second R', 'r1', 0.5, 0.1, 10.0, 0.01)
+        self.sliders.add_slider(
+            'Time of change', 't1', mid_point, bounds[0], bounds[1], 1)
+
+        new_rs = [2.0, 0.5]
+        start_times = [0, mid_point]
+        simulator.model.set_r_profile(new_rs, start_times)
+
+        data = simulator.run(10)
+        df = pd.DataFrame({
+            'Time': simulator.get_regime(),
+            'Incidence Number': data})
+
+        self.plot.add_simulation(df)
+
+        self.simulator = simulator
+
+    def get_sliders_ids(self):
+        """
+        Returns the IDs aof all sliders accompaning the figure in the
+        app.
+        """
+        return self.sliders.slider_ids()
+
+    def update_simulation(self, new_init_cond, new_r0, new_r1, new_t1):
+        """
+        Updates the model parameters in the simulator and the
+        simulated graph in the figure.
+        """
+        new_rs = [new_r0, new_r1]
+        start_times = [0, new_t1]
+
+        model = self.simulator.model
+        model.set_r_profile(new_rs, start_times)
+
+        data = self.simulator.run(new_init_cond)
+        fig = self.plot.figure
+        fig['data'][1]['y'] = data
```

Hardcoding the index 1 here is causing problems. Depending on when you called add_simulator relative to add_data, the correct index could vary.

You could fix this by saving the correct index, or alternatively the correct plotly graph object, as an instance variable at the end of add_simulator. Then, use that information in this method. 

If you tried this, you'd probably also want to raise an error if update_simulation gets called before add_simulator.